### DISCUSSION TITLE
Git-LFS custom-transfer error in special circumstance: windows, self-built, local fetch

### DISCUSSION THREAD

**Discussion by olsen232** [git-lfs/git-lfs#5196] [Q&A] [answered]
Ok so I'm experimenting with git-lfs as part of some work in [kart](https://github.com/koordinates/kart).
At this stage, I'm building a git-lfs binary and then calling it from some python code to do some low level git-lfs operations.
During testing, I set up two git repositories locally, and then try and use `git-lfs fetch` to copy an LFS object from one to the other. 

This works on linux and macos. However, on windows, it fails with the following error message:
```
git-lfs fetch origin d312bb50d08e095b7fe8709e0cd7e2a5f804fe9b
fetch: Fetching reference d312bb50d08e095b7fe8709e0cd7e2a5f804fe9b
failed to find custom transfer command "git-lfs" remote: executable file not found in %PATH%
error: failed to fetch some objects from 'file:///D:/Users/aolsen/mykart/olsen-pc-23-nov/.git'
```

More things to know: the custom build wasn't supposed to be very custom - it's just built from git-lfs at master using `go build`. 
I don't understand the error message. I wasn't trying to do a custom transfer (unless maybe I am if this is necessary to fetch things from a file:/// repository?). The custom git-lfs.exe *is* on the path, so I'm not really sure what it's complaining about.
The git-lfs.exe that I built works well otherwise - it's good at doing "single-player" commands locally eg `git-lfs ls-files`, and it's good at fetching objects from an LFS server. It's just when it starts trying to talk to itself to communicate between two repos that it stops working.

Have I just built it wrong, or is there something else that's going on here?

**Comment by chrisd8088** [comment, marked as answer]
Hey, I'm sorry you're having trouble.  I'm not sure I can give you a precise answer, only some possible ideas.

> I wasn't trying to do a custom transfer (unless maybe I am if this is necessary to fetch things from a file:/// repository?)

I believe that's the case for `file://` transfers, yes, per this code [comment](https://github.com/git-lfs/git-lfs/blob/941b616aaf6f4adfa176616dc70c8d5e01b4a3d1/lfshttp/client.go#L116-L121) and the manual [page](https://github.com/git-lfs/git-lfs/blob/941b616aaf6f4adfa176616dc70c8d5e01b4a3d1/docs/man/git-lfs-standalone-file.adoc) for the `git-lfs-standalone-file(1)` command.  So I think the Git LFS client is calling itself after [constructing](https://github.com/git-lfs/git-lfs/blob/941b616aaf6f4adfa176616dc70c8d5e01b4a3d1/tq/custom.go#L357) a custom transfer adapter using the command filename `git-lfs` and the command arguments `standalone-file`, i.e., it's expecting to be able to run `git-lfs standalone-file`, which should in turn run the `git-lfs-standalone-file` binary.

Now as for why, on Windows, that doesn't locate your `git-lfs.exe`, I can only speculate a bit.  First, Git LFS deliberately doesn't look in the current directory (i.e., `.`) for executables on Windows, even though Windows itself will do so.  You may therefore need to put an explicit path to your custom-built binary into your `PATH` environment variable.

Second, in order to match the `.exe` part of the actual binary's filename, you would want to make sure the `PATHEXT` environment variable has that file extension listed, if it is defined at all.  If it's not defined at all, either as a system or a user environment variable, we should [default](https://github.com/git-lfs/git-lfs/blob/941b616aaf6f4adfa176616dc70c8d5e01b4a3d1/subprocess/path_windows.go#L89) to a list that includes `.exe`, but if that variable is defined and non-empty, for instance, it would override that default.

**Reply by chrisd8088** [reply]
There's also the difference in Windows between the user and system `PATH` environment variables to be conscious of; I can't recall offhand how those are merged to provide a set of paths to running processes.  Since you're building your own `git-lfs` binary from source, you might want to insert a tracing output statement right after the [call](https://github.com/git-lfs/git-lfs/blob/941b616aaf6f4adfa176616dc70c8d5e01b4a3d1/subprocess/path.go#L53) to `os.Getenv("PATH")` to see what it contains.  If you `import` the `"github.com/rubyist/tracerx"` package you can just call something like `tracerx.Printf("PATH: %s", path)` at that point in the code, and then running with `GIT_TRACE=1` in your environment should cause that line, among others, to be output when you run `git-lfs`.

You could also add some tracing output around this [loop](https://github.com/git-lfs/git-lfs/blob/941b616aaf6f4adfa176616dc70c8d5e01b4a3d1/subprocess/path.go#L64-L67) to see what paths and file extensions are being searched, and why your binary isn't found.

**Comment by olsen232** [comment]
Thanks for the diagnosis tip - I added the tracing to git-lfs as you suggested. The problem in the end was that there was no `sh.exe` anywhere on the PATH, which I guess is also required for the `git-lfs-standalone-file` flow. Making sure that `sh.exe` was somewhere on the path solved the problem.
